core/trino-main/src/main/java/io/trino/operator/window/matcher/ThreadEquivalence.java

```diff
@@ -185,24 +251,64 @@ private static int resolvePosition(LogicalIndexNavigation navigation, ArrayView
     }
 
-    private static List<List<PhysicalValuePointer>> getClassifierValuePointers(List<List<PhysicalValuePointer>> valuePointers)
+    private static List<List<PhysicalValuePointer>> getClassifierValuePointers(List<List<PhysicalValueAccessor>> valuePointers)
     {
         return valuePointers.stream()
                 .map(pointerList -> pointerList.stream()
+                        .filter(pointer -> pointer instanceof PhysicalValuePointer)
+                        .map(PhysicalValuePointer.class::cast)
                         .filter(pointer -> pointer.getSourceChannel() == CLASSIFIER)
                         .collect(toImmutableList()))
                 .collect(toImmutableList());
     }
 
+    // for every label, iterate over aggregations in the label's defining condition, and divide them into sublists:
+    // - aggregations which do not depend on `CLASSIFIER`, that is, either do not use `CLASSIFIER` in any of their arguments,
+    // or apply to only one label, in which case the result of `CLASSIFIER` is always the same.
+    // - aggregations which depend on `CLASSIFIER`, that is, use `CLASSIFIER` in some of their arguments,
+    // and apply to more than one label (incl. the universal pattern variable), in which case the result of `CLASSIFIER`
+    // depends on the actual matched label.
+    private static AggregationIndexes classifyAggregations(List<List<PhysicalValueAccessor>> valuePointers, List<MatchAggregationLabelDependency> labelDependencies)
+    {
+        ImmutableList.Builder<List<Integer>> noClassifierAggregations = ImmutableList.builder();
+        boolean foundNoClassifierAggregations = false;
+        ImmutableList.Builder<List<Integer>> classifierAggregations = ImmutableList.builder();
+        boolean foundClassifierAggregations = false;
+
+        for (List<PhysicalValueAccessor> pointerList : valuePointers) {
+            ImmutableList.Builder<Integer> noClassifierAggregationIndexes = ImmutableList.builder();
+            ImmutableList.Builder<Integer> classifierAggregationIndexes = ImmutableList.builder();
+            for (PhysicalValueAccessor pointer : pointerList) {
+                if (pointer instanceof MatchAggregationPointer) {
+                    int aggregationIndex = ((MatchAggregationPointer) pointer).getIndex();
+                    MatchAggregationLabelDependency labelDependency = labelDependencies.get(aggregationIndex);
+                    if (!labelDependency.isClassifierInvolved() || labelDependency.getLabels().size() == 1) {
+                        foundNoClassifierAggregations = true;
+                        noClassifierAggregationIndexes.add(aggregationIndex);
+                    }
+                    else {
+                        foundClassifierAggregations = true;
+                        classifierAggregationIndexes.add(aggregationIndex);
+                    }
+                }
+            }
+            noClassifierAggregations.add(noClassifierAggregationIndexes.build());
```

Comment:
It means that for a given label, in the label's defining condition, there are no aggregations of this kind:
- "no-classifier aggregations" is the kind of aggregations whose result does not depend on the actual labels, e.g. `max(A.price)`
- "classifier aggregations"  is the kind of aggregations whose result depends on the actual labels, e.g. `array_agg(CLASSIFIER(U))`, where `U` is a subset `{A, B}`

These two kinds of aggregations are explained in the comment above the method `classifyAggregations`.

This might look like arbitrary partitioning, but it is based on the observation that for some aggregations it is enough to compare the sets of positions, while other aggregations require also comparing the matched labels.

The overall purpose is to optimize `ThreadEquivalence` and make it more accurate by skipping unnecessary comparisons.